src/main/ml-modules/root/test/test-controller.xqy

```diff
@@ -71,23 +71,67 @@ declare function list()
 		}
 };
 
-declare function run-suite($suite as xs:string, $tests as xs:string*, $run-suite-teardown as xs:boolean, $run-teardown as xs:boolean) {
+(:~
+: Execute all of the test suites and report with code coverage information
+:)
+declare function run-tests() {
+	let $run-suite-teardown := fn:true()
+	let $run-teardown := fn:true()
+	let $calculate-coverage := fn:true()
+	let $test-results :=
+		element t:tests {
+			for $suite in list()/t:suite
+			let $suite-name as xs:string := $suite/@path
+			let $test-modules as xs:string* := $suite/t:tests/t:test/@path
+			return run-suite($suite-name, $test-modules, $run-suite-teardown, $run-teardown, $calculate-coverage)
+		}
+	return
+		cover:summary($test-results)
+};
+
+(:~
+ : Execute the suite tests with the options specified. Default behavior is not to calculate code coverage.
+ :)
+declare function run-suite(
+		$suite as xs:string,
+		$tests as xs:string*,
+		$run-suite-teardown as xs:boolean,
+		$run-teardown as xs:boolean)
+{
+		run-suite($suite, $tests, $run-suite-teardown, $run-teardown, fn:false())
+};
+
+(:~
+ : Execute the suite tests with the options specified.
+ :)
+declare function run-suite(
+		$suite as xs:string,
+		$tests as xs:string*,
+		$run-suite-teardown as xs:boolean,
+		$run-teardown as xs:boolean,
+		$calculate-coverage as xs:boolean)
+{
 	let $start-time := xdmp:elapsed-time()
+	let $tests as xs:string* :=
+		if ($tests) then $tests
+		else list()/t:suite[@path eq $suite]/t:tests/t:test/@path
+	let $coverage :=
+		if ($calculate-coverage) then
+			(: TODO: should we exclude the test modules from what is covered?
+					i.e. cover:list-coverage-modules()[fn:not(fn:starts-with(., $TEST-SUITES-ROOT))]
+			:)
+			cover:prepare(cover:list-coverage-modules(), $tests ! fn:concat($TEST-SUITES-ROOT, $suite, "/", .))
+		else ()
 	let $results :=
 		element t:run {
 			helper:log(" "),
 			helper:log(text {"SUITE:", $suite}),
 			run-setup-teardown(fn:true(), $suite),
 
 			helper:log(" - invoking tests"),
-
-			let $tests as xs:string* :=
-				if ($tests) then $tests
-				else
-					list()/t:suite[@path = $suite]/t:tests/t:test/@path
 			for $test in $tests
 			return
-				run($suite, $test, fn:concat($TEST-SUITES-ROOT, $suite, "/", $test), $run-teardown),
+        run($suite, $test, fn:concat($TEST-SUITES-ROOT, $suite, "/", $test), $run-teardown, $coverage),
```

Comment:
Please use tabs to match the whitespace in the rest of this file.

Reply:
Sorry about that. I've gotten into the habit of using the 2 space convention, but should have kept with existing indentation. If the convention is tabs in this project, would you prefer that I apply tabs for the newly added modules as well?

Reply:
I prefer two spaces as well, but I try to suppress that in favor of whatever's being used in the file. I'm planning to add a .editorconfig file that will make it easier to keep this straight. For the other files, yeah, sticking with whatever's prevalent in the project makes sense.

Reply:
@hansenmc it looks like there are only a couple files in this project that use tabs. I'd keep your new files with 2 spaces.